parsl/executors/status_handling.py

```diff
@@ -199,15 +199,32 @@ def scale_out(self, blocks: int = 1) -> List[str]:
                 self._simulated_status[block_id] = JobStatus(JobState.FAILED, "Failed to start block {}: {}".format(block_id, ex))
         return block_ids
 
-    @abstractmethod
     def scale_in(self, blocks: int) -> List[str]:
         """Scale in method.
 
         Cause the executor to reduce the number of blocks by count.
 
+        The default implementation will kill blocks without regard to their
+        status or whether they are executing tasks. Executors with more
+        nuanced scaling strategies might overload this method to work with
+        that strategy - see the HighThroughputExecutor for an example of that.
+
         :return: A list of block ids corresponding to the blocks that were removed.
         """
-        pass
+        # Obtain list of blocks to kill
+        to_kill = list(self.blocks_to_job_id.keys())[:blocks]
+        kill_ids = [self.blocks_to_job_id[block] for block in to_kill]
+
+        # Cancel the blocks provisioned
+        if self.provider:
+            logger.info(f"Scaling in jobs: {kill_ids}")
+            r = self.provider.cancel(kill_ids)
```

Comment:
- in the error case, no sense in defining `kill_ids`
- if it reads clearer to you (does to me) `kill_ids` could be inlined:

    ```python
    if self.provider:
        kill_ids = [v for i, v in enumerate(self.blocks_to_job_id.values()) if i < blocks]
        logger.info(f"Scaling in jobs: {kill_ids}")
        ...
    ```
- On the other hand, I see that this mostly consolidating already-existing code, and the PR description mentions fixing it in a follow-up.  Your call.

Reply:
yeah this PR i really wanted to just do copy/paste refactoring.

Looking at this again (not for this PR):

I also think the else path should properly be treated as an error:

either raise an exception, or be an `assert` instead of an `if` with the rule that scale_in should only be called from internal code, and that code shouldn't be running if there is no provider.